xdsl/dialects/gpu.py

```diff
@@ -354,10 +355,7 @@ def verify_(self) -> None:
 class ModuleEndOp(IRDLOperation):
     name = "gpu.module_end"
 
-    # TODO circular dependency disallows this set of traits
-    # tracked by gh issues https://github.com/xdslproject/xdsl/issues/1218
-    # traits = frozenset([HasParent(ModuleOp), IsTerminator()])
-    traits = frozenset([IsTerminator()])
+    traits = traits_def(lambda: frozenset([IsTerminator()]))
```

Comment:
I think the `HasParent` trait got removed (maybe accidentally?).
AFAIU it is still [needed](
https://mlir.llvm.org/docs/Dialects/GPU/#gpumodule_end-gpumoduleendop).
BTW, the correct definition was in gh tracking issue comment that was removed.

Unless there is another reason for the commission that I don't see.